### PR TITLE
Better typeck divergence

### PR DESCRIPTION
**Pull request by varkor**
This pulls in some of the ideas from https://github.com/rust-lang/rust/pull/47291 about divergence checking as a first step, namely: if a function takes arguments of an uninhabited type, it must be diverging, and it is often possible to check for uninhabited types rather than the never type specifically.

There are still a few places special-cased on `is_never`, which I left alone because I didn't want to mess around with the special coercions involving `!`, but I might have been over-cautious.

This adds a new case, `Diverges::UnwarnedAlways`, that specifically does not lint for `unreachable_code`, because I'm conscious it's going to break anything involving (for example) empty enum impls. If we're happy doing that, though, this case can be removed.

(Arguably some/most of the changes involving `is_never` are not typeck related, but they seemed in a similar vein, so I included them in this PR. I can split if necessary though.)

r? @arielb1 
cc @nikomatsakis @canndrew

### CODE REVIEW

**Comment by arielb1** [comment]
I would prefer at least a lang team FCP for these sorts of changes. cc @rust-lang/lang

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -607,7 +607,12 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
 
         // If there are no arms, that is a diverging match; a special case.
         if arms.is_empty() {
-            self.diverges.set(self.diverges.get() | Diverges::Always);
+            let diverges = match self.diverges.get() {
```
Comment (arielb1) [inline src/librustc_typeck/check/_match.rs]: Why does this convert `UnwarnedAlways` to `Always`? This looks like something that needs to be clarified.

Comment (varkor) [reply src/librustc_typeck/check/_match.rs]: At the moment, this is special casing to keep the warning behaviour the same as before. I would prefer not to special-case, but I wanted to know what the feelings were about changing the unreachable code warnings.

**Comment by varkor** [comment]
I'll explicate what's going on in a comment on `Diverges`. So far, the only thing that changes with regards to user-facing warnings is that tuple types are considered uninhabited if all their components are. If possible, I'd prefer to start emitting a few more warnings (places that have been special-cased around `UnwarnedAlways` at the moment).

> Also, as implemented, this makes some code legal and insta-stable, so this probably requires some stability gating.

Do you think it'd be better to make the changes to the warnings, but feature gate them instead then? I'll do that for now!

**Comment by arielb1** [comment]
The test is testing something that is *definitely* UB:
https://github.com/rust-lang/rust/blob/3c43aa5677e17dfb9f5004522745a02e21a604a4/src/test/debuginfo/nil-enum.rs#L30-L43

It should be changed to not be UB.

**Comment by varkor** [comment]
I couldn't see any reason to keep that test around, as its entire purpose was UB as far as I can tell. I've added more of a description to `Diverges` as to its behaviour, but I've left off adding a new stability gate until a decision is made.

**Comment by arielb1** [comment]
@varkor 

I do want to have a test that checks that the debug-info is sane (e.g., does not kill gdb).

For example, you could do the following sort of test:
```Rust
enum Empty {}

fn zzz() {}

fn main() {
    let x: *const Empty = 1 as *const _;

    zzz();
}
```

**Comment by nox** [comment]
#50304 has some overlap, I didn't care about type checking at all but I used the proper layout infra for the uninhabitedness check in trans.

File: src/test/debuginfo/nil-enum.rs [hunk 2]
```diff
@@ -15,31 +15,21 @@
 // compile-flags:-g
 // gdb-command:run
 
-// gdb-command:print first
```
Comment (arielb1) [inline src/test/debuginfo/nil-enum.rs]: Why did you remove the `gdbr-check`?

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/_match.rs]: is this comment out of date now?

File: src/libsyntax/feature_gate.rs [hunk 3]
```diff
@@ -466,6 +466,9 @@ declare_features! (
 
     // #[doc(alias = "...")]
     (active, doc_alias, "1.27.0", Some(50146), None),
+
+    // Enables improved divegence checking in functions/match arms.
```
Comment (nikomatsakis) [inline src/libsyntax/feature_gate.rs]: Nit: dive*r*gence

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (varkor) [reply src/librustc_typeck/check/_match.rs]: Yes: I replaced `UnwarnedAlways` with the feature gating.

File: src/librustc_mir/build/expr/into.rs [hunk 4]
```diff
@@ -213,8 +213,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 exit_block.unit()
             }
             ExprKind::Call { ty, fun, args } => {
-                // FIXME(canndrew): This is_never should probably be an is_uninhabited
-                let diverges = expr.ty.is_never();
+                let diverges = expr.ty.conservative_is_uninhabited();
```
Comment (nikomatsakis) [inline src/librustc_mir/build/expr/into.rs]: seems safe 👍

File: src/librustc/cfg/construct.rs [hunk 5]
```diff
@@ -415,8 +415,7 @@ impl<'a, 'tcx> CFGBuilder<'a, 'tcx> {
             args: I) -> CFGIndex {
         let func_or_rcvr_exit = self.expr(func_or_rcvr, pred);
         let ret = self.straightline(call_expr, func_or_rcvr_exit, args);
-        // FIXME(canndrew): This is_never should probably be an is_uninhabited.
-        if self.tables.expr_ty(call_expr).is_never() {
+        if self.tables.expr_ty(call_expr).conservative_is_uninhabited() {
```
Comment (nikomatsakis) [inline src/librustc/cfg/construct.rs]: *probably* safe 👍

File: src/librustc/ty/sty.rs [hunk 6]
```diff
@@ -1288,6 +1288,27 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
...
 
+    pub fn conservative_is_uninhabited(&self) -> bool {
+        // Uncontentious uninhabitableness check
+        match self.sty {
+            ty::TyNever => true,
+            // We could extend this to ADTs in which every variant is uninhabited
+            // (for enums, unions, etc.).
+            ty::TyAdt(def, _) => def.variants.is_empty(),
+            ty::TyTuple(tys) => tys.iter().any(|ty| ty.conservative_is_uninhabited()),
+            ty::TyArray(ty, len) => {
```
Comment (nikomatsakis) [inline src/librustc/ty/sty.rs]: Any particular reason to expand the set of cases here? Not that I I necessarily object. It seems though that including `(!,)` but not `struct Foo(!)` is sort of surprising to me -- in other words, I'm tempted to stick to `!` and `Void`, and separate out this aspect of the PR.

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs [hunk 7]
```diff
@@ -986,8 +986,7 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                 }
             }
             None => {
-                // FIXME(canndrew): This is_never should probably be an is_uninhabited
-                if !sig.output().is_never() {
+                if !sig.output().conservative_is_uninhabited() {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/type_check/mod.rs]: seems ok 👍

File: src/librustc_mir/hair/pattern/check_match.rs [hunk 8]
```diff
@@ -225,7 +225,15 @@ impl<'a, 'tcx> MatchVisitor<'a, 'tcx> {
                 let scrutinee_is_uninhabited = if self.tcx.features().exhaustive_patterns {
                     self.tcx.is_ty_uninhabited_from(module, pat_ty)
                 } else {
-                    self.conservative_is_uninhabited(pat_ty)
+                    if self.tcx.features().better_divergence_checking {
```
Comment (nikomatsakis) [inline src/librustc_mir/hair/pattern/check_match.rs]: Maybe we should lump this feature gate in with `exhaustive_patterns`? I'm not sure yet that I see the difference here

File: src/librustc_trans/debuginfo/mod.rs [hunk 9]
```diff
@@ -271,7 +271,7 @@ pub fn create_function_debug_context<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>,
         }
         None => {}
     };
-    if sig.output().is_never() {
+    if sig.output().conservative_is_uninhabited() {
```
Comment (nikomatsakis) [inline src/librustc_trans/debuginfo/mod.rs]: seems ok 👍

File: src/librustc_trans/declare.rs [hunk 10]
```diff
@@ -133,8 +133,7 @@ pub fn declare_fn<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>, name: &str,
     let fty = FnType::new(cx, sig, &[]);
     let llfn = declare_raw_fn(cx, name, fty.llvm_cconv(), fty.llvm_type(cx));
 
-    // FIXME(canndrew): This is_never should really be an is_uninhabited
-    if sig.output().is_never() {
+    if sig.output().conservative_is_uninhabited() {
```
Comment (nikomatsakis) [inline src/librustc_trans/declare.rs]: seems ok 👍

File: src/librustc_typeck/check/mod.rs [hunk 11]
```diff
@@ -433,6 +433,14 @@ pub enum PlaceOp {
 /// wake). Tracked semi-automatically (through type variables marked
 /// as diverging), with some manual adjustments for control-flow
 /// primitives (approximating a CFG).
+/// We know a node diverges in the following (conservative) situations:
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: Nit: new paragraph please

File: src/librustc_typeck/check/mod.rs [hunk 12]
```diff
@@ -1064,6 +1072,15 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
...
             ty::BindingMode::BindByValue(hir::Mutability::MutImmutable), true);
 
+        // If any of a function's parameters have a type that is uninhabited, then it
+        // may never be called (because its arguments cannot be constructed). Therefore,
+        // it must always diverge.
+        if fcx.tcx.features().better_divergence_checking {
+            if arg_ty.conservative_is_uninhabited() {
+                fcx.diverges.set(fcx.diverges.get() | Diverges::Always);
+            }
+        }
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: it's definitely an interesting "judgement call" whether we want to issue a warning in this case

Comment (nikomatsakis) [reply src/librustc_typeck/check/mod.rs]: (probably?)

File: src/librustc_typeck/check/mod.rs [hunk 13]
```diff
@@ -3538,9 +3555,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             _ => self.warn_if_unreachable(expr.id, expr.span, "expression")
         }
 
-        // Any expression that produces a value of type `!` must have diverged
-        if ty.is_never() {
-            self.diverges.set(self.diverges.get() | Diverges::Always);
+        // Any expression that produces a value of an uninhabited type must have diverged.
+        if ty.conservative_is_uninhabited() {
+            if ty.is_never() || self.tcx.features().better_divergence_checking {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: seems good 👍 -- in particular, seems like a natural extension of the existing check, and it's not clear that `!` should be special here

**Review by nikomatsakis — COMMENTED** [review]
So one of the things that was a bit tricky with this PR for me was that I felt like it's goals were not clearly articulated. Here are the things I think it does, let me know if I am missing anything:

- Extend the "conservative" definition of uninhabited to include not only `!` and empty enums but also tuples and non-empty arrays of said types.
  - `&!` is still not included
  - Neither is `struct Foo(!)`
- Expands the check for functions that return `!` to include any uninhabited type
  - this also affects liveness, debuginfo, and a few other random places
- Marks function bodies as dead code whenever function parameters are uninhabited
  - maybe we should special case an empty body here for the purposes of warnings?
  - I believe an empty body would always be legal, because of the rule that a block with an implied tail expression can be coerced to `!` -- or perhaps we could make that true anyway?

I *think* I'm in favor of all 3 of those goals, though the first one seems somewhat separable to me from the rest. That part makes me the most nervous, though I think I cannot imagine any unsafe code guidelines that would permit `(!,)` (and indeed I think this is already UB in existing code).

File: src/librustc_mir/hair/pattern/check_match.rs (hunk 8 above)
Comment (varkor) [reply src/librustc_mir/hair/pattern/check_match.rs]: Well, the one reason it might be worth separating is because the change to uninhabited function parameter warning may not really be well suited to `exhaustive_patterns`. But it does seem like overkill to have another feature for this, so I'll add it there for now.

**Comment by varkor** [comment]
> So one of the things that was a bit tricky with this PR for me was that I felt like it's goals were not clearly articulated.

Yeah, I did amalgamate some separate but related issues here. I'll separate the first one out into a follow-up.

> Here are the things I think it does

Yes, those are right.

> - maybe we should special case an empty body here for the purposes of warnings?
> - I believe an empty body would always be legal, because of the rule that a block with an implied tail expression can be coerced to !

Perhaps — this would require some tweaks, as (despite [1]) empty blocks don't seem to be coerced even when diverging (granted, this possibility is new as of this PR, so that's probably not unexpected). I'll look into doing that.

[1] https://github.com/rust-lang/rust/blob/986cc4d1402c95ea399581f74ac303b7cfd654a6/src/librustc_typeck/check/mod.rs#L4415-L4422

**Comment by varkor** [comment]
> maybe we should special case an empty body here for the purposes of warnings?

Okay, so now empty bodies will not warn, and neither will functions with _just_ the trailing expression (as this also is a common pattern for impls of uninhabited types). This should make the lint less annoying when a function is necessarily uninhabited.

**Comment by pietroalbini** [comment]
Ping from triage @arielb1! This PR needs your review.

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @arielb1 or someone else from @rust-lang/lang review this?

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @arielb1 or someone else from @rust-lang/compiler @rust-lang/lang review this?

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @arielb1 or someone else from @rust-lang/compiler @rust-lang/lang review this?

**Comment by pietroalbini** [comment]
Ping from triage @rust-lang/lang @rust-lang/compiler! This PR has been ignored for more than three weeks, can someone check in on it?

**Comment by TimNN** [comment]
Another ping,  @rust-lang/lang / @rust-lang/compiler could one of you please comment on how to move forward with this PR or add this to your teems meeting agenda?

**Comment by pietroalbini** [comment]
Moved to waiting-on-team. @rust-lang/compiler @rust-lang/lang this PR is waiting for a review for more than a month.

**Comment by pietroalbini** [comment]
Ping from triage @nikomatsakis! This PR needs your review.

File: src/librustc_typeck/check/mod.rs [hunk 14]
```diff
@@ -1066,6 +1066,13 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
         fcx.check_pat_walk(&arg.pat, arg_ty,
             ty::BindingMode::BindByValue(hir::Mutability::MutImmutable), true);
 
+        // If any of a function's parameters have a type that is uninhabited, then it
+        // may never be called (because its arguments cannot be constructed). Therefore,
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: Nit: I would say "then it can never be called". "may never" sounds to me like it *might not* be called, but it still *could be*

File: src/librustc_typeck/check/mod.rs [hunk 15]
```diff
@@ -3557,9 +3557,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-        if ty.is_never() {
-            self.diverges.set(self.diverges.get() | Diverges::Always);
+        // Any expression that produces a value of an uninhabited type must have diverged.
+        if ty.conservative_is_uninhabited() {
+            let always = if ty.is_never() {
+                Diverges::Always
+            } else {
+                // We avoid linting in this case for
+                // consistency with previous versions.
+                Diverges::UnwarnedAlways
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: I have a question here -- if we later process an expression that produces a value of type `!`, would we lint then? e.g., `foo(a, b)` where `a: &!` and `b: !`.

Comment (nikomatsakis) [reply src/librustc_typeck/check/mod.rs]: I seem to recall reading that `UnwarnedAlways` can never be converted to `WarnedAlways`... so I am guessing not? Or maybe I am misinterpreting/misremembering how that enum works.

**Review by nikomatsakis — COMMENTED** [review]
Argh I didn't get as far as I'd like in this review, so leaving some "in progress" comments here.

File: src/librustc_typeck/check/mod.rs (hunk 15 above)
Comment (varkor) [reply src/librustc_typeck/check/mod.rs]: Hmm, yes, I suppose it wouldn't lint a later unreachable expression, which probably isn't what we want. I'll check that.

**Comment by stokhos** [comment]
Ping from triage @varkor , any progress on this PR?

**Comment by varkor** [comment]
@stokhos: this is still waiting on @nikomatsakis. I'll address all the comments at once, as that's going to be easier to review.

**Comment by arielb1** [comment]
@stokhos 

I think this is still waiting on the lang-team decision on how to proceed.

File: src/librustc_mir/hair/pattern/check_match.rs [hunk 16]
```diff
@@ -225,7 +225,15 @@ impl<'a, 'tcx> MatchVisitor<'a, 'tcx> {
                 let scrutinee_is_uninhabited = if self.tcx.features().exhaustive_patterns {
                     self.tcx.is_ty_uninhabited_from(module, pat_ty)
                 } else {
-                    self.conservative_is_uninhabited(pat_ty)
+                    if self.tcx.features().exhaustive_patterns {
+                        pat_ty.conservative_is_uninhabited()
```
Comment (arielb1) [inline src/librustc_mir/hair/pattern/check_match.rs]: Isn't this checked 2 lines before, preventing this from being reached?

Comment (arielb1) [reply src/librustc_mir/hair/pattern/check_match.rs]: i.e. `match` uses `is_ty_uninhabited_from` instead of `conservative_is_uninhabited`.

**Comment by stokhos** [comment]
Ping from triage! @varkor any update on this PR?

**Comment by varkor** [comment]
I'll fix the comments pointed out so far, soon, but this is still waiting on a final review.

File: src/librustc_mir/hair/pattern/check_match.rs (hunk 16 above)
Comment (varkor) [reply src/librustc_mir/hair/pattern/check_match.rs]: Ah, this is where using `exhaustive_patterns` for two different things causes problems. I'll default to the old check here instead.

File: src/librustc_typeck/check/mod.rs (hunk 15 above)
Comment (varkor) [reply src/librustc_typeck/check/mod.rs]: Actually, this shouldn't be an issue, because the only time we use `Diverges::UnwarnedAlways` is with an empty function that can't be executed, but we don't want to trigger the lint for (as in https://github.com/rust-lang/rust/pull/50262#pullrequestreview-116439632). In such a case, no more expressions are available to trigger a divergence warning.

**Comment by varkor** [comment]
I've addressed all the current comments.

**Comment by pietroalbini** [comment]
Ping from triage @nikomatsakis! This PR needs your review.

**Comment by TimNN** [comment]
Ping from triage @nikomatsakis  / @rust-lang/compiler. This PR requires your review.

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @nikomatsakis or someone else from @rust-lang/compiler review this?

**Comment by Centril** [comment]
We decided on the lang team meeting to revisit this next week.

File: src/test/compile-fail/unreachable-try-pattern.rs [hunk 17]
```diff
@@ -18,6 +18,7 @@ enum Void {}
 impl From<Void> for i32 {
     fn from(v: Void) -> i32 {
         match v {}
+        //~^ WARN unreachable expression
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/test/compile-fail/unreachable-try-pattern.rs]: If not this code, what would you expect the user to write?

File: src/test/compile-fail/unreachable-try-pattern.rs [hunk 18]
```diff
@@ -39,6 +40,7 @@ fn qux(x: Result<u32, Void>) -> Result<u32, i32> {
 fn vom(x: Result<u32, Void>) -> Result<u32, i32> {
     let y = (match x { Ok(n) => Ok(n), Err(e) => Err(e) })?;
     //~^ WARN unreachable pattern
+    //~| WARN unreachable expression
```
Comment (nikomatsakis) [inline src/test/compile-fail/unreachable-try-pattern.rs]: Similarly here, it feels unfortunate to start warning unless we have a nicer way to write this code (but, at present, we don't...?).

If we had the `!` pattern concept, we could perhaps do 

```rust
match x { 
    Ok(n) => Ok(n),
    Err(!)
}
```

File: src/test/debuginfo/nil-enum.rs [hunk 19]
```diff
@@ -15,31 +15,20 @@
...
+// gdb-command:print *first
 // gdbg-check:$1 = {<No data fields>}
 // gdbr-check:$1 = <error reading variable>
 
-// gdb-command:print second
```
Comment (nikomatsakis) [inline src/test/debuginfo/nil-enum.rs]: Why these diffs?

Comment (nikomatsakis) [reply src/test/debuginfo/nil-enum.rs]: I guess because the code as it was written is basically bogus...

File: src/test/ui/better_divergence_checking.rs [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+    // A function with an uninhabited parameter
+    // is permitted if its body is empty.
+}
+
+pub fn uninhabited_parameter_ii(v: !) -> i32 {
+    // A function with an uninhabited parameter
+    // is permitted if it simply returns a value
+    // as a trailing expression to satisfy the
+    // return type.
+    v
```
Comment (nikomatsakis) [inline src/test/ui/better_divergence_checking.rs]: I guess this answers my question to "how should the user write this code" -- instead of `match void { }` they would simply write `void`?

But [that will only work if the uninhabited parameter has type `!`...?](https://play.rust-lang.org/?gist=154ed08d778627693a4cef2d889860ec&version=nightly&mode=debug&edition=2015) Or, maybe we changed that in this PR somewhere?

File: src/test/ui/feature-gate-better_divergence_checking.rs [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+// except according to those terms.
+
+pub enum Void {}
+
+fn main() {
+    let x: (Void,) = unsafe { std::mem::uninitialized() };
+    let _ = match x {}; //~ ERROR non-exhaustive
+
+    let x: [Void; 1] = unsafe { std::mem::uninitialized() };
+    let _ = match x {}; //~ ERROR non-exhaustive
```
Comment (nikomatsakis) [inline src/test/ui/feature-gate-better_divergence_checking.rs]: I think if we have a feature-gate test, we should also have an equivalent test but with the feature-gate enabled so we can see the difference — is there such a test somewhere? (If so, I missed it in the PR)

File: src/librustc_mir/hair/pattern/check_match.rs [hunk 22]
```diff
@@ -227,7 +227,11 @@ impl<'a, 'tcx> MatchVisitor<'a, 'tcx> {
                 let scrutinee_is_uninhabited = if self.tcx.features().exhaustive_patterns {
                     self.tcx.is_ty_uninhabited_from(module, pat_ty)
                 } else {
-                    self.conservative_is_uninhabited(pat_ty)
+                    match pat_ty.sty {
+                        ty::TyNever => true,
+                        ty::TyAdt(def, _) => def.variants.is_empty(),
+                        _ => false
+                    }
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_mir/hair/pattern/check_match.rs]: I think it would be identical?

**Review by nikomatsakis — COMMENTED** [review]
OK, I did another pass. I think my feeling on this PR is that I am vaguely 👍 on the direction it is going but not sure about the point that it reaches. I'm a bit nervous, for example, about the new warnings — and I'd like to know what our overall plan is.

I am wondering, @varkor, if you'd be interested in starting work on an RFC, perhaps incorporating the [ideas around a `!` pattern](http://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/), to kind of "set the plan" here?

I'd be interested in seeing such an RFC and helping to discuss it, but I don't think I can take lead on it.

File: src/librustc_typeck/check/mod.rs [hunk 23]
```diff
@@ -1047,6 +1064,24 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
...
+        // If any of a function's parameters have a type that is uninhabited, then it
+        // it is not possible to call that function (because its arguments cannot be constructed).
+        // Therefore, it must always diverge.
+        if fcx.tcx.features().exhaustive_patterns {
+            if arg_ty.conservative_is_uninhabited() {
+                let mut diverges = Diverges::Always;
+                if let hir::ExprKind::Block(ref block, _) = body.value.node {
+                    // If the function is completely empty, or has a single trailing
+                    // expression, then we do not issue a warning (as it was likely
+                    // mandated by a trait, rather than being an oversight).
```
[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc_typeck/check/mod.rs]: hmm... did you mean to write something else in the second snippet? They are identical...

**Comment by nikomatsakis** [comment]
Removing nomination.

**Comment by varkor** [comment]
I'm going to write a little summary now and think about how to approach this more comprehensively.

This PR was intended to increase the number of cases taken into account by the typeck divergence checking. This divergence checking is just used for error messages (e.g. `unreachable code`).

One problem here is that it's being done on an expression level, which for uninhabited arguments arguably isn't the right place to warn, which is leading to most of the issues with the warning message changes here.

The other problem is that although some warnings technically apply to uninhabited types in general, because Rust only coerces `!` to any other type, "useless code" is actually necessary, so the warnings, though accurate, are unavoidable.

I think the correct solution for now is to split the changes in this PR up and change the warning for uninhabited parameter types to a new warning that's more informative.
